### PR TITLE
CORE-9567 Manual Approval docs

### PR DESCRIPTION
**Pull request by YashNabar**

### CODE REVIEW

File: wiki/Manual-Approval.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Manual approval of registration requests
+
+Membership groups in Corda 5 may be configured to approve (or decline) member registration requests manually. The MGM
+operator can configure the approval method for joining the group, so that requests satisfying pre-defined criteria would
+require manual approval, while others would be auto-approved. Manual approval presents the request to the MGM operator,
+and requires the operator to validate the request through an off-line procedure before approving/declining it via HTTP
+RPC. This applies to registration and re-registration requests alike. The configuration may be added at any point in
+time, and only affects future registration requests - previously approved members will not be required to re-register.
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: It would be best not to add a new line in the middle of sentences/paragraphs so that it can be rendered according to the screen size and editor configuration of the reader. Same for the next paragraph.

```suggestion
Membership groups in Corda 5 may be configured to approve (or decline) member registration requests manually. The MGM operator can configure the approval method for joining the group, so that requests satisfying pre-defined criteria would require manual approval, while others would be auto-approved. Manual approval presents the request to the MGM operator, and requires the operator to validate the request through an off-line procedure before approving/declining it via HTTP RPC. This applies to registration and re-registration requests alike. The configuration may be added at any point in time, and only affects future registration requests - previously approved members will not be required to re-register.
```

File: wiki/Manual-Approval.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
+# Manual approval of registration requests
+
+Membership groups in Corda 5 may be configured to approve (or decline) member registration requests manually. The MGM
+operator can configure the approval method for joining the group, so that requests satisfying pre-defined criteria would
+require manual approval, while others would be auto-approved. Manual approval presents the request to the MGM operator,
+and requires the operator to validate the request through an off-line procedure before approving/declining it via HTTP
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: I would maybe reword this slightly. They can validate the request through an offline procedure, but it's not required. It's up to the operator.

```suggestion
and allows the operator to review the request before approving/declining it via the REST API.
```

File: wiki/Manual-Approval.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+RPC. This applies to registration and re-registration requests alike. The configuration may be added at any point in
+time, and only affects future registration requests - previously approved members will not be required to re-register.
+
+# How to configure a group for manual approval
+
+Registration requests are evaluated according to regular expression-based rules submitted by the MGM operator. The
+proposed `MemberInfo` is compared with the previous (if any) `MemberInfo` to calculate the difference. This difference
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]:
```suggestion
proposed `MemberInfo` is compared with the previous (if any) `MemberInfo` to calculate the difference in their member contexts. This difference
```

File: wiki/Manual-Approval.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+the MGM. If any of the keys present in this `MemberInfo` difference match the regular expressions set by the MGM
+operator, the request will require manual approval. If there are no matches, the request is auto-approved.
+
+**Add a group approval rule**
+
+<details>
+<summary>Bash</summary>
+
+```bash
+RULE_PARAMS='{"ruleParams":{"ruleRegex": "corda.*", "ruleLabel": "All Corda"}}'
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: nit-picky but maybe something like this will be clearer for the reader

```suggestion
RULE_PARAMS='{"ruleParams":{"ruleRegex": "corda.*", "ruleLabel": "Review all changes to keys in the Corda namespace"}}'
```

File: wiki/Manual-Approval.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+```bash
+curl --insecure -u admin:admin -X GET $API_URL/mgm/$MGM_HOLDING_ID/approval/rules
+```
+
+</details>
+
+**Delete a group approval rule**
+
+Replace `<RULE ID>` with the ID of the rule to be deleted.
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]:
```suggestion
Replace `<RULE ID>` with the ID of the rule to be deleted. The rule ID can be retrieve from the response of creating a rule, or from the response of the GET endpoint described previously.
```

File: wiki/Manual-Approval.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+Requests which are pending manual approval are assigned `PENDING_MANUAL_APPROVAL` status.
+
+**View requests pending manual approval**
+
+<details>
+<summary>Bash</summary>
+
+```bash
+curl --insecure -u admin:admin -X 'GET' https://localhost:8888/api/v1/mgm/$MGM_HOLDING_ID/registrations
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: Does this command work OK for you without `?viewinactive=false`?
I just remembered I had an issue while writing the e2e tests where this endpoint failed without `viewinactive` even though it should be optional.

File: wiki/Manual-Approval.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+</details>
+
+To optionally view requests from a specific member (e.g. C=GB, L=London, O=Alice) and/or include historic requests, use:
+
+<details>
+<summary>Bash</summary>
+
+```bash
+curl --insecure -u admin:admin -X 'GET' 'https://localhost:8888/api/v1/mgm/$MGM_HOLDING_ID/registrations?requestsubjectx500name=C%3DGB%2C%20L%3DLondon%2C%20O%3DAlice&viewhistoric=true'
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: Nit-picky, I'm sure this works, but for consistency it would be good to either use `'` on all methods in the commands on this page or none. None would be my preference. 
```suggestion
curl --insecure -u admin:admin -X GET 'https://localhost:8888/api/v1/mgm/$MGM_HOLDING_ID/registrations?requestsubjectx500name=C%3DGB%2C%20L%3DLondon%2C%20O%3DAlice&viewhistoric=true'
```

Comment (charlieR3) [reply wiki/Manual-Approval.md]: I think `'` is only used here and the one previous.

File: wiki/Manual-Approval.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+</details>
+
+**View current group approval rules**
+
+<details>
+<summary>Bash</summary>
+
+```bash
+curl --insecure -u admin:admin -X GET $API_URL/mgm/$MGM_HOLDING_ID/approval/rules
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: I think `-X GET` might actually be the default if you don't specify the method so you could simplify this here and in other commands I think to 
```suggestion
curl --insecure -u admin:admin $API_URL/mgm/$MGM_HOLDING_ID/approval/rules
```

File: wiki/Manual-Approval.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+> Note: This only works with requests that are in `PENDING_MANUAL_APPROVAL` status.
+
+Replace `<REQUEST ID>` with the ID of the registration request.
+
+<details>
+<summary>Bash</summary>
+
+```bash
+REQUEST_ID=<REQUEST ID>
+curl --insecure -u admin:admin -X POST -d "" $API_URL/mgm/$MGM_HOLDING_ID/approve/$REQUEST_ID
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: I think if you use `-d` it will automatically use the `POST` method so no need for `-X POST`. Although since it's an empty body I think you could omit `-d` if `-X POST` is included so maybe this
```suggestion
curl --insecure -u admin:admin -X POST $API_URL/mgm/$MGM_HOLDING_ID/approve/$REQUEST_ID
```

File: wiki/Manual-Approval.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+> Note: This only works with requests that are in `PENDING_MANUAL_APPROVAL` status.
+
+Replace `<REQUEST ID>` with the ID of the registration request.
+
+<details>
+<summary>Bash</summary>
+
+```bash
+REQUEST_ID=<REQUEST ID>
+REASON='{"reason":{"reason": "test"}}'
```
Comment (charlieR3) [inline wiki/Manual-Approval.md:98]: Pretty sure you're already aware but this will need to be changed after my PR gets merged.

Comment (charlieR3) [reply wiki/Manual-Approval.md:98]: But I can do that if this is merged first. I'll need to update this page for pre-auth approval also

File: wiki/Manual-Approval.md (hunk 6 above)
Comment (YashNabar) [reply wiki/Manual-Approval.md]: It worked fine for me when I tested this manually as part of CORE-9059

Comment (charlieR3) [reply wiki/Manual-Approval.md]: Ok, that's good then. Not sure why the test failed on it then.

**Review by charlieR3 — APPROVED** [review]

File: wiki/Manual-Approval.md (hunk 10 above)
Comment (YashNabar) [reply wiki/Manual-Approval.md:98]: If you want, I can add the pre-auth part here (including this payload format update)

File: wiki/Manual-Approval.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+
+# How to configure a group for manual approval
+
+Registration requests are evaluated according to regular expression-based rules submitted by the MGM operator. The proposed `MemberInfo` is compared with the previous (if any) `MemberInfo` to calculate the difference in their member contexts. This difference will be 100% in case of a first-time registration, since there will be no previous `MemberInfo` for that member known to the MGM. If any of the keys present in this `MemberInfo` difference match the regular expressions set by the MGM operator, the request will require manual approval. If there are no matches, the request is auto-approved.
+
+**Add a group approval rule**
```
Comment (charlieR3) [inline wiki/Manual-Approval.md]: Last comment, but would these be better as subheaders? e.g.
```suggestion
## Add a group approval rule
```

File: wiki/Manual-Approval.md (hunk 10 above)
Comment (charlieR3) [reply wiki/Manual-Approval.md:98]: Do you mean the description of creating pre-auth tokens and rules? I actually started that yesterday so I don't mind finishing it.

Comment (charlieR3) [reply wiki/Manual-Approval.md:98]: I'll add it to the bottom of this doc

**Comment by YashNabar** [comment]
Added to the Wiki - https://github.com/corda/corda-runtime-os/wiki/Manual-Approval
